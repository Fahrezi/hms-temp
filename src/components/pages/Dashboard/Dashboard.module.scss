.dashboard {
  & {
    padding: 2rem;
  }

  &__title {
    font-size: 2rem;
    font-weight: 600;
    margin-bottom: 2rem;
  }

  &__content {
    display: grid;
    gap: 1.5rem;
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  }

  &__card {
    background: white;
    border-radius: 0.5rem;
    padding: 1.5rem;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.1);

    &--heading {
      font-size: 1.25rem;
      font-weight: 500;
      margin-bottom: 1rem;
    }

    &--paragraph {
      color: #666;
    }
  }

  &__image {
    margin-top: 100px;
  }
}
