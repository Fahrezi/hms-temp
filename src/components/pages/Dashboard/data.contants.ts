// === Raw data from your old dashboard ===
export const TOTAL_ROOMS = 344;

// Room status
export const VD = 148;
export const VC = 45;
export const VCI = 23;
export const OD = 118;
export const OC = 1;
export const OOO_TEMP = 0;
export const OOO_PERM = 0;
export const HU_TEMP = 5;
export const HU_PERM = 0;
export const HU_PRMO = 0;
export const COMP_GCOMP = 0;
export const COMP_FOC = 2;


// Top-line
export const FOC_OCHU = 90; // OC-HU & FOC
export const EST_OCC_PERCENT = 26.16; // %


// Room movement (Rooms)
export const Rooms = {
  lastNight: 159,
  departed: 41,
  arrived: 1,
  expectedArrivals: 16,
  expectedDepartures: 38,
  estWalkIn: 0,
  estOccRooms: 97,
};


// Guest movement (Pax)
export const Pax = {
  lastNight: 293,
  departed: 80,
  arrived: 2,
  expectedArrivals: 34,
  expectedDepartures: 75,
  estWalkIn: 0,
  estOccPax: 174,
};


// Derived
export const OCCUPIED_NOW = OC + OD; // current occupied by HK status snapshot
export const VACANT_CLEAN_READY = VC + VCI; // ready inventory
export const INTERNAL_OR_BLOCKED = HU_TEMP + HU_PERM + HU_PRMO + COMP_FOC + OOO_TEMP + OOO_PERM; // internal use / maintenance
export const classifiedSum = OCCUPIED_NOW + VACANT_CLEAN_READY + VD + INTERNAL_OR_BLOCKED;
export const OTHER = Math.max(TOTAL_ROOMS - classifiedSum, 0); // any residual to keep sums consistent


export const occupancyBreakdown = [
  { name: "Occupied (OC+OD)", value: OCCUPIED_NOW },
  { name: "Vacant Clean/Inspected (VC+VCI)", value: VACANT_CLEAN_READY },
  { name: "Vacant Dirty (VD)", value: VD },
  { name: "Internal / OOS (HU/COMP/OOO)", value: INTERNAL_OR_BLOCKED },
  { name: "Other/Unclassified", value: OTHER },
];


export const COLORS = ["#2563eb", "#22c55e", "#ef4444", "#f59e0b", "#64748b"]; // blue, green, red, amber, slate


export const flowData = [
  { label: "Arrivals", value: Rooms.arrived },
  { label: "Expected Arrivals", value: Rooms.expectedArrivals },
  { label: "Departures", value: Rooms.departed },
  { label: "Expected Departures", value: Rooms.expectedDepartures },
  { label: "Est. Walk-In", value: Rooms.estWalkIn },
];


export const roomsPaxTable = [
  {
    row: "Last Night",
    rooms: Rooms.lastNight,
    pax: Pax.lastNight,
  },
  {
    row: "Departed",
    rooms: Rooms.departed,
    pax: Pax.departed,
  },
  {
    row: "Arrived",
    rooms: Rooms.arrived,
    pax: Pax.arrived,
  },
  {
    row: "Expected Arrival",
    rooms: Rooms.expectedArrivals,
    pax: Pax.expectedArrivals,
  },
  {
    row: "Expected Departure",
    rooms: Rooms.expectedDepartures,
    pax: Pax.expectedDepartures,
  },
  {
    row: "Estimated Occupied (EoD)",
    rooms: Rooms.estOccRooms,
    pax: Pax.estOccPax,
  },
];

export const roomAvailability = [
  {
    "type_of_room": "Standard Room",
    "availability": [12, 10, 8, 7, 9, 11, 13]
  },
  {
    "type_of_room": "Deluxe Room",
    "availability": [8, 7, 6, 5, 6, 7, 8]
  },
  {
    "type_of_room": "Suite Room",
    "availability": [3, 3, 2, 2, 2, 3, 3]
  },
  {
    "type_of_room": "Family Room",
    "availability": [5, 5, 4, 4, 5, 5, 6]
  },
  {
    "type_of_room": "Executive Room",
    "availability": [4, 4, 3, 3, 4, 4, 5]
  },
  {
    "type_of_room": "Presidential Suite",
    "availability": [1, 1, 1, 0, 1, 1, 1]
  }
];

export const optionsAvailability = [
  { label: 'GTD/Confirmed', value: 'confirmed' },
  { label: 'Include Tentative', value: 'tentative' },
  { label: 'Include Waiting List', value: 'waiting_list' },
];

export const statusOptionList = {
  traces: {
    label: "Traces",
    data: [
      { "label": "Message", "value": "message" },
      { "label": "Complaint", "value": "complaint" },
      { "label": "Location", "value": "location" },
      { "label": "Wake Up Call", "value": "wake_up_call" },
      { "label": "Reminder", "value": "reminder" },
      { "label": "Inventory Request", "value": "inventory_request" },
      { "label": "Leisure Booking", "value": "leisure_booking" }
    ]
  },
  lostAndFound: {
    label: 'Lost & Found',
    data: [
      { "label": "Lost & Fund", "value": "lost_and_found" },
    ]
  },
  room: {
    label: 'Room',
    data:[
      { "label": "Queue", "value": "room_queue" },
    ]
  },
  specialDate: {
    label: 'Special Date',
    data: [
      { "label": "New Year’s Day", "value": "2025-01-01" },
      { "label": "Valentine’s Day", "value": "2025-02-14" },
      { "label": "Easter Sunday", "value": "2025-04-20" },
      { "label": "Labor Day", "value": "2025-05-01" },
      { "label": "Independence Day", "value": "2025-07-04" },
      { "label": "Halloween", "value": "2025-10-31" },
      { "label": "Christmas Eve", "value": "2025-12-24" },
      { "label": "Christmas Day", "value": "2025-12-25" },
      { "label": "New Year’s Eve", "value": "2025-12-31" }
    ]
  },
  guestType: {
    label: 'Guest Type',
    data: [
      { "label": "Reguler", "value": "guest_reguler" },
      { "label": "VIP", "value": "guest_vip" },
    ]
  },
  guestEntitlement: {
    label: 'Guest Entitlement',
    data: [
      { "label": "VIP Status Accorded", "value": "vip_status_accorded" },
      { "label": "Standard Cake", "value": "standard_cake" },
      { "label": "Free 1 Time Kayak 30 Min", "value": "free_1_time_kayak_30_min" },
      { "label": "Suite / IMV Benefit", "value": "suite_imv_benefit" },
      { "label": "Standard Fruit Basket", "value": "standard_fruit_basket" },
      { "label": "Private Transfer Upon", "value": "private_transfer_upon" },
      { "label": "Free 2Soft Drink vcr the", "value": "free_2soft_drink_vcr_the" }
    ]
  },
  guestHighlight: {
    label: 'Guest Highlight',
    data: []
  },
  group: {
    label: 'Group',
    data: []
  },
  promotion: {
    label: 'Promotion',
    data: []
  },
  miceVenue: {
    label: 'M.I.C.E Venue',
    data: [
      { "label": "Ballroom", "value": "ballroom" },
      { "label": "Grand Ballroom", "value": "grand_ballroom" },
    ]
  },
  miceFunction: {
    label: 'M.I.C.E Function',
    data: []
  },
  forex: {
    label: 'Forex',
    data: [
      { "label": "IDR", "value": "10000" },
      { "label": "SGD", "value": "100" },
    ]
  }
}
