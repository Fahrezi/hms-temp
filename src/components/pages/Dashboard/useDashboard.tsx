import { useLocation } from 'react-router-dom';

import { getUsers } from '@/services/user.service';

import { User, UsersResponse } from '@/types/user';

const useDashboard = () => {
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);
  const pageFromUrl: number = Number(searchParams.get('page'));

  const { data, isFetching } = getUsers(pageFromUrl);

  const responseDataUser: UsersResponse | undefined = data?.data;
  const usersData: User[] = responseDataUser?.data ?? [];
  const totalPages: number = responseDataUser?.total_pages ?? 0;

  return {
    usersData,
    totalPages,
    isFetching,
  };
};

export default useDashboard;
