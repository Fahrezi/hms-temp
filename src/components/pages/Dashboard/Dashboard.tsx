import { useNavigate } from 'react-router-dom';

import LazyImage from '@/components/ui/LazyImage';
import Loading from '@/components/ui/Loading';
import Pagination from '@/components/ui/Pagination';

import useDashboard from './useDashboard';

import styles from './Dashboard.module.scss';

const Dashboard = () => {
  const { usersData, totalPages, isFetching } = useDashboard();
  const navigate = useNavigate();

  return (
    <div className={styles.dashboard}>
      <h1 className={styles.dashboard__title}>Dashboard</h1>
      <div className={styles.dashboard__content}>
        <div className={styles.dashboard__card}>
          <h2 className={styles['dashboard__card--heading']}>Welcome</h2>
          <p className={styles['dashboard__card--paragraph']}>This is your dashboard</p>
        </div>
      </div>
      <br />
      {isFetching ? (
        <Loading />
      ) : (
        <div>
          {usersData.map((item, key) => (
            <h3
              onClick={() => {
                void navigate(`/detail/${item.user_id}`);
              }}
              key={key}
            >
              {item.email}
            </h3>
          ))}
          <br />
        </div>
      )}
      <br />
      <br />
      {usersData && <Pagination maxPage={totalPages} />}
      <LazyImage src='/images/hero.jpg' alt='hero' className={styles.dashboard__image} />
    </div>
  );
};

export default Dashboard;
