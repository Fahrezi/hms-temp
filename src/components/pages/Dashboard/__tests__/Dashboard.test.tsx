/* eslint-disable sonarjs/no-duplicate-string */
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { vi } from 'vitest';

import Dashboard from '@/components/pages/Dashboard';
import useDashboard from '@/components/pages/Dashboard/useDashboard';

// Mock hooks
vi.mock('@/components/pages/Dashboard/useDashboard', () => ({
  default: vi.fn(),
}));

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: vi.fn(),
  };
});

const mockUseDashboard = useDashboard as unknown as ReturnType<typeof vi.fn>;

describe('Dashboard Component', () => {
  const mockNavigate = vi.fn();

  beforeEach(() => {
    vi.mocked(useNavigate).mockReturnValue(mockNavigate);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders the snapshot Dashboard component', () => {
    mockUseDashboard.mockReturnValue({
      usersData: [],
      totalPages: 0,
      isFetching: true,
    });

    const { container } = render(
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders loading state when data is fetching', () => {
    mockUseDashboard.mockReturnValue({
      usersData: [],
      totalPages: 0,
      isFetching: true,
    });

    render(
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>,
    );

    expect(screen.getByTestId('loading')).toBeInTheDocument();
  });

  it('renders user data and pagination when fetching is complete', () => {
    mockUseDashboard.mockReturnValue({
      usersData: [
        {
          user_id: 1,
          email: 'user1@example.com',
          display_name: '',
          roles: [],
        },
        {
          user_id: 2,
          email: 'user2@example.com',
          display_name: '',
          roles: [],
        },
      ],
      totalPages: 2,
      isFetching: false,
    });

    render(
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>,
    );

    expect(screen.getByText('user1@example.com')).toBeInTheDocument();
    expect(screen.getByText('user2@example.com')).toBeInTheDocument();
    expect(screen.getByText('Dashboard')).toBeInTheDocument();
  });

  it('navigates to user detail page on click', () => {
    mockUseDashboard.mockReturnValue({
      usersData: [
        {
          user_id: 1,
          email: 'user1@example.com',
          display_name: '',
          roles: [],
        },
      ],
      totalPages: 1,
      isFetching: false,
    });

    render(
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>,
    );

    const userItem = screen.getByText('user1@example.com');
    fireEvent.click(userItem);

    expect(mockNavigate).toHaveBeenCalledWith('/detail/1');
  });
});
