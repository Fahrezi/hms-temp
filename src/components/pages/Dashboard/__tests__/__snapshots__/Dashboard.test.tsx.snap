// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dashboard Component > renders the snapshot Dashboard component 1`] = `
<div>
  <div
    class="_dashboard_22c2a6"
  >
    <h1
      class="_dashboard__title_22c2a6"
    >
      Dashboard
    </h1>
    <div
      class="_dashboard__content_22c2a6"
    >
      <div
        class="_dashboard__card_22c2a6"
      >
        <h2
          class="_dashboard__card--heading_22c2a6"
        >
          Welcome
        </h2>
        <p
          class="_dashboard__card--paragraph_22c2a6"
        >
          This is your dashboard
        </p>
      </div>
    </div>
    <br />
    <svg
      class=""
      data-testid="loading"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <style>
        
          .spinner {
            transform-origin: center;
            animation: rotate 0.75s infinite linear;
          }
          @keyframes rotate {
            100% { transform: rotate(360deg); }
          }
        
      </style>
      <path
        d="M12,1A11,11,0,1,0,23,12,11,11,0,0,0,12,1Zm0,19a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
        opacity="0.25"
      />
      <path
        d="M12,4a8,8,0,0,1,7.89,6.7A1.53,1.53,0,0,0,21.38,12h0a1.5,1.5,0,0,0,1.48-1.75,11,11,0,0,0-21.72,0A1.5,1.5,0,0,0,2.62,12h0a1.53,1.53,0,0,0,1.49-1.3A8,8,0,0,1,12,4Z"
      />
    </svg>
    <br />
    <br />
    <div />
    <div
      class="_imageContainer_d76e9f _dashboard__image_22c2a6"
      style="width: 100%; aspect-ratio: 16/9;"
    >
      <div
        class="_placeholder_d76e9f"
      />
    </div>
  </div>
</div>
`;
